.github/copilot-instructions.md

```diff
@@ -0,0 +1,137 @@
+- Remove nodriver (CI requirement): `pip uninstall -y nodriver`
+- Install in editable mode: `pip install -e .` -- takes 30 seconds. NEVER CANCEL.
+
+### Testing and Validation
+- Run unit tests: `python -m etc.unittest` -- takes 3-5 seconds. NEVER CANCEL.
+- Expected results: ~41 tests, 1-2 failures expected (network isolation), 5-8 skipped tests
+- Time unit tests: `time python -m etc.unittest` for precise timing
+- Individual test scripts available in `etc/testing/` but may have outdated model references
+
+### Running the Application
+- **CLI Help**: `g4f --help` or `python -m g4f --help`
+- **Client CLI**: `g4f client --help` for interactive text generation
+- **API Server**: `python -m g4f --port 8080` -- starts FastAPI server with web GUI on http://localhost:8080
+- **API Server Only**: `python -m g4f --port 8080 --no-gui` (if supported)
+- **Python Library**: 
+  ```python
+  from g4f.client import Client
+  client = Client()
+  response = client.chat.completions.create(model="gpt-4o", messages=[{"role": "user", "content": "Hello"}])
+  print(response.choices[0].message.content)
+  ```
+
+## Validation Scenarios
+
+Always test these scenarios after making changes:
+
+### Basic Functionality Test
+1. **Import Test**: `python -c "from g4f.client import Client; client = Client(); print('Success')"`
+2. **CLI Test**: `g4f --help` should show available commands without errors
+3. **Unit Tests**: `python -m etc.unittest` should complete without hanging
+4. **API Server Test**: Start server `python -m g4f --port 8080`, then `curl -I http://localhost:8080` should return 200 OK
+
+### Manual Validation Requirements
+- ALWAYS test complete user workflows, not just imports
+- Test both CLI and Python library interfaces when making provider changes
+- Test API server startup and basic HTTP response
+- Verify no new import errors or dependency conflicts
+
+## Critical Timing and Build Information
+
+- **NEVER CANCEL** long-running operations. Builds may take several minutes.
+- **Requirements Installation**: 
+  - Minimal (`requirements-min.txt`): 30-60 seconds. Set timeout to 120+ seconds.
+  - Full (`requirements.txt`): 2-5 minutes. Set timeout to 600+ seconds.
+- **Unit Tests**: 3-5 seconds. Set timeout to 30+ seconds.
+- **Package Installation**: 30-60 seconds. Set timeout to 120+ seconds.
+- **Warning**: `pydub` will show ffmpeg warning - this is expected and harmless
+
+## Key Development Info
+
+### Project Structure
+- `g4f/` - Main library code
+  - `Provider/` - AI provider implementations (108+ providers)
+  - `client/` - Client interfaces
+  - `api/` - FastAPI server implementation
+  - `gui/` - Web GUI components
+  - `cli/` - Command line interfaces
+- `etc/` - Testing and development tools
+  - `unittest/` - Unit test suite
+  - `testing/` - Integration tests (may have outdated references)
+  - `examples/` - Usage examples
+- `docs/` - Documentation
```

Comment:
You can find the documentation here:

https://github.com/gpt4free/g4f.dev/tree/main/docs